xla/service/scatter_determinism_expander.cc

```diff
@@ -0,0 +1,459 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/scatter_determinism_expander.h"
+#include <cstdint>
+#include "absl/container/flat_hash_set.h"
+
+#include "tsl/platform/logging.h"
+#include "tsl/platform/statusor.h"
+#include "xla/hlo/ir/hlo_casting_utils.h"
+#include "xla/hlo/ir/hlo_computation.h"
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_instructions.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/hlo/ir/hlo_opcode.h"
+#include "xla/literal_util.h"
+#include "xla/service/hlo_creation_utils.h"
+#include "xla/shape_util.h"
+#include "xla/xla_data.pb.h"
+#include "xla/service/scatter_utils.h"
+
+namespace xla {
+
+// Canonicalizes the scatter_updates in order to keep them uniform while
+// performing the scatter operation.
+static StatusOr<std::vector<HloInstruction*>> CanonicalizeScatterUpdates(
+    const std::vector<HloInstruction*>& scatter_updates,
+    HloInstruction* scatter_indices, const ScatterDimensionNumbers& dim_numbers,
+    int64_t scatter_loop_trip_count) {
+  std::vector<HloInstruction*> adjusted_updates;
+  adjusted_updates.reserve(scatter_updates.size());
+  for (HloInstruction* update : scatter_updates) {
+    TF_ASSIGN_OR_RETURN(
+        HloInstruction * canonical_update,
+        PermuteScatterAndWindowDims(update, dim_numbers.update_window_dims()));
+    TF_ASSIGN_OR_RETURN(
+        HloInstruction * adjusted_update,
+        AdjustScatterDims(scatter_indices->shape(), canonical_update,
+                          dim_numbers.index_vector_dim()));
+    CHECK_EQ(scatter_loop_trip_count, adjusted_update->shape().dimensions(0));
+    adjusted_updates.push_back(adjusted_update);
+  }
+  return adjusted_updates;
+}
+
+// Create the out-of-bound tensor for the scatter operation.
+HloInstruction* CreateOutOfBoundTensor(HloComputation* parent,
+                                       HloInstruction* scatter_indices,
+                                       const Shape& scatter_shape) {
+  if (scatter_indices->shape().rank() == 1) {
+    CHECK(scatter_shape.dimensions_size() == 1);
```

Comment:
nit: CHECK_EQ instead of CHECK() with '==' op

Reply:
changed